### ISSUE TITLE
The config should be persisted across app starts

### ISSUE THREAD

**Issue by gus-byrnesinnovation** [launchdarkly/ios-client#9]
Currently when the app comes down the config is lost until the next connection to the API. We should persist the config and load it from the data store upon init of the library.

**Comment by jeff-byrnesinnovation** [comment]
@jkodumal it sounds like you and @gus-byrnesinnovation changed how you want the config to be persisted, since what we discussed before I left. In the past it was my belief, that is there was a Config stored, we would use it no matter what. Are the following scenarios correct?

1) If the developer passes a User object, and the key matches the persisted User, then we update the User object and use the persisted Config
2) If the developer passes a User object, and the key does not match the persisted User, then we use the default values
3) If the developer passes a User object, and there is no persisted User, then we use the default values

**Comment by jkodumal** [comment]
I don't think this matches what Gus and I discussed in e-mail. Reiterating what I wrote in e-mail:

My suggestion would be to scope the persisted flag configuration data by user key. I think this has some nice properties:

1) If you tweak the attributes on a user, but the user is intended to be the same (key doesn’t change), then they’ll receive the old flag values until the new ones are polled.

2) If you switch users completely then you’ll wait to fetch a fresh configuration if none exists in the store, or use the last persisted if the key has been seen before.

The only downside to this would be the potentially unbounded number of flag configurations stored. Solutions:

1) Bound the number of flag configurations to say 5, with an LRU policy.

2) Accept that it is unlikely that a large number of users will be accessing an app on a single device.

**Comment by jkodumal** [comment]
Conceptually, my suggestion is to store the flag configuration data as a map:

user key -> flag key -> flag value

Your last comment suggests that there is one persisted user. I'm suggesting that there is a persisted config per user key.

**Comment by jeff-byrnesinnovation** [comment]
No, we save more than one User that we look up by key.

User -> Config -> Flag Key -> Flag Value

So the only difference with what you said is that you want to limit it to 5 stored Users?

**Comment by gus-byrnesinnovation** [comment]
Jeff, John

Hold off on merging my pull request. I just realized I left the user
restriction out and realized that the user is not updated currently but a
new record is created. I will need to add one more method to the
DataManager to find and update vs find or create. I will fix the last part
tonight and update the pull.

Gus

**Comment by jeff-byrnesinnovation** [comment]
@jkodumal 

Gus and I are working on persisting the config with the logic of the LRU cache for 5 Users. Based on our current implementation, if the developer does not pass the key in for the User, then that User and its config will be persisted but never be able to be retrieved on subsequent starts of the client. We just wanted to make sure that matches what you desired.

Thanks,
Jeff

**Comment by jkodumal** [comment]
If the developer were to save the generated key, and set up their user with
that same key later, they would be able to access the persisted data,
right? I don't think anyone would do that-- just trying to understand the
implementation.

**Comment by cmavromoustakos** [comment]
John

That is not the current implementation. Currently we are not persisting if
the user is not passing in a key. The thought process for leaving it out
was that if a user just used the key that we assigned for them they would
have to save it in some persistence layer themselves to reference it on
subsequent app starts. If you think this is a valid case we can add it to
that one as well.

Again sorry this is spanning over 2 weeks. When I implemented the user
persistence originally I started finding edge cases (multiple users/configs
being stored) based on the way we implemented the original user and config
lookup. We had implemented with the expectation that there can be only one.
So we have been bouncing things back and forth between Jeff and myself to
make sure we catch all these edge cases.

Gus
